pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linuxrouting
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
+
+	"github.com/vishvananda/netlink"
+)
+
+// MigrateENIRoutingTable migrates the egress rules inside the Linux routing
+// policy database (rpdb) for ENI IPAM mode.
+//
+// We rely on the ability to fetch the CiliumNode resource because we need to
+// fetch the number associated with the ENI device. The CiliumNode resource
+// contains this information in the Status field.
+//
+// This function should be invoked before exposing the Cilium API because we
+// want to ensure that no workloads are scheduled while this modification is
+// taking place. This migration is related to a bug
+// (https://github.com/cilium/cilium/issues/14336) where an ENI has an ifindex
+// that equals the main routing table number (253-255), causing the rules and
+// routes to be created using the wrong table ID, which could end up
+// blackholing most traffic on the node.
+func MigrateENIRoutingTable(getter interfaceDB) []error {
+	var errors []error
+
+	rules, err := netlink.RuleList(netlink.FAMILY_V4)
+	if err != nil {
+		err = fmt.Errorf("failed to list rules, cannot migrate ENI routing table: %w", err)
+		errors = append(errors, err)
+		return errors
+	}
+
+	for _, r := range rules {
```

Comment:
Let's split this entire loop into two phases:

Phase 1: Set up the new tables and rules. This short be done in a first iteration of all rules. If this fails , we should cleanup and Fatal. THe system should look the same as before.

Phase 2: Delete the old rules to swap over to the new model. If this fails, we should re-add the deleted rules and then Fatal as well.